src/discrete.jl

```diff
@@ -0,0 +1,121 @@
+using StaticArrays, ForwardDiff, DiffEqBase
+using DiffEqBase: isinplace
+using OrdinaryDiffEq: FunctionMap
+import DiffEqBase: init, step!, isinplace
+
+export MinimalDiscreteProblem, MinimalDiscreteIntegrator
+#####################################################################################
+#                          Minimal Discrete Problem                                 #
+#####################################################################################
+"""
+    MinimalDiscreteProblem(eom, state, p = nothing, t0 = 0)
+"""
+struct MinimalDiscreteProblem{IIP, F, S, P, D, T} <: DEProblem
+    # D, T are dimension and eltype of state
+    f::F      # eom, but same syntax as ODEProblem
+    u0::S     # initial state
+    p::P      # parameter container
+    t0::Int   # initial time
+end
+MDP = MinimalDiscreteProblem
+function MinimalDiscreteProblem(eom::F, state, p::P = nothing, t0 = 0) where {F, P}
+    IIP = isinplace(eom, 4)
+    # Ensure that there are only 2 cases: OOP with SVector or IIP with Vector
+    # (requirement from ChaosTools)
+    IIP || @assert typeof(eom(state, p, 0)) <: SVector
+    u0 = IIP ? Vector(state) : SVector{length(state)}(state...)
+    S = typeof(u0)
+    D = length(u0); T = eltype(u0)
+    MinimalDiscreteProblem{IIP, F, S, P, D, T}(eom, u0, p, t0)
+end
+isinplace(::MDP{IIP}) where {IIP} = IIP
+mutable struct MinimalDiscreteIntegrator{IIP, F, S, P, D, T} <: DEIntegrator
+    prob::MDP{IIP, F, S, P, D, T}
+    u::S      # integrator state
+    t::Int    # integrator "time" (counter)
+    dummy::S  # dummy, used only in the IIP version
+    p::P      # parameter container, I don't know why
+end
+MDI = MinimalDiscreteIntegrator
+isinplace(::MDI{IIP}) where {IIP} = IIP
+
+function init(prob::MDP{IIP, F, S, P, D, T},
+    u = prob.u0) where {IIP, F, S, P, D, T}
+    u0 = IIP ? Vector(u) : SVector{length(u)}(u...)
+    return MDI{IIP, F, S, P, D, T}(prob, u0, prob.t0, deepcopy(u0), prob.p)
+end
+
+#####################################################################################
+#                                   Stepping                                        #
+#####################################################################################
+# IIP version
+function step!(integ::MDI{true})
+    integ.dummy .= integ.u

```

Comment:
You could just do a pointer swap.

Reply:
I'll need help here. I don't know what pointers are :D (besides "conceptually")

Reply:
Sounds like... you need a pointer. Okay okay, I couldn't resist!